src/puppetlabs/puppetdb/jdbc.clj

```diff
@@ -127,17 +127,43 @@
    #(and (map? %)
          (valid-jdbc-query? (:results-query %)))))
 
-(defn convert-result-arrays
-  "Converts Java and JDBC arrays in a result set using the provided function
-  (eg. vec, set). Values which aren't arrays are unchanged."
-  ([result-set]
-     (convert-result-arrays vec result-set))
-  ([f result-set]
-     (let [convert #(cond
-                     (kitchensink/array? %) (f %)
-                     (isa? (class %) java.sql.Array) (f (.getArray %))
-                     :else %)]
-       (map #(kitchensink/mapvals convert %) result-set))))
+(defn convert-if-array
+  ([x] (convert-if-array vec x))
+  ([convert x]
+   (cond
+     (kitchensink/array? x) (convert x)
+     (isa? (class x) java.sql.Array) (convert (.getArray x))
+     :else x)))
+
+(defn call-with-query-rows
+  "Calls (f rows), where rows is a lazy sequence of rows generated
+  from within a transaction.  Converts any java.sql.Array
+  or (.isArray (class v)) values to a vector.  The sequence is backed
+  by an active database cursor which will be closed when f returns.
+  Cancels the query if f throws an exception.  The option names that
+  correspond to jdbc/result-set-seq options will affect the rows
+  produced as they do for that function.  So for example, when
+  as-arrays? is logically true, the result rows will be vectors, not
+  maps, and the first result row will be a vector of column names."
+  ([query f] (call-with-query-rows query {} f))
+  ([[sql & params]
+    {:keys [as-arrays? identifiers qualifier read-columns] :as opts}
+    f]
+   (with-db-transaction []
+     (with-open [stmt (.prepareStatement (:connection *db*) sql)]
+       (doseq [[index value] (map-indexed vector params)]
```

Comment:
Hmm, I'll take a look -- what's there now is just what was already in with-query-results-cursor, which this is replacing.